devito/finite_differences/operations.py

```diff
@@ -0,0 +1,164 @@
+import sympy
+
+import devito
+from devito.symbolics.search import retrieve_functions
+from devito.symbolics.extended_sympy import FrozenExpr
+
+__all__ = ['Mul', 'Add', 'Pow']
+
+
+def fd_parameters(obj):
+    """
+    Process a set of Mul, Add, Pow expression and outputs
+    the space_order, time_order of the whole expression.
+    As the leading order is always the smallest order, if multiple
+    orders are found the lowest is chosen
+    """
+    func = list(retrieve_functions(obj))
+    time_dims = ()
+    sp_dims = ()
+    so = 100
+    to = 100
+    is_time = False
+    stagg = (0, 0, 0, 0)
+    # Filter expressions to get space and time Order
+    # The space order is the minimum space order
+    # The time order is the minimum time_order
+    for i in func:
+        if isinstance(i, devito.Function):
```

Comment:
i.is_Function

Reply:
This would fail, the only reason this weird loop is that in some case we end up with like `cos(theta)` which doesn't have a `is_Function`

Reply:
then we have a bug in `retrieve_functions`, as I seem to remember it is only expected to fetch our own Function/TimeFunction.

Random guess. It's possible the bug is due to ``cos/sin/...`` being SymPy object on which ``is_Function == True`` (as these are indeed..."true" callable functions!). If that's the case, maybe we could fix ``retrieve_functions`` by looking for devito.Functions (``isinstance(...)``), rather than resorting to ``is_Function``.

Reply:
Ok I'll look at it